src/scheduler/scheduler.rs

```diff
@@ -80,36 +85,61 @@ impl<VM: VMBinding> GCWorkScheduler<VM> {
         self.worker_group.as_ref().worker_count()
     }
 
-    /// Create GC threads, including the controller thread and all workers.
+    /// Create GC threads for the first time.  It will also create the `GCWorker` instances.
+    ///
+    /// Currently GC threads only include worker threads, and we currently have only one worker
+    /// group.  We may add more worker groups in the future.
     pub fn spawn_gc_threads(self: &Arc<Self>, mmtk: &'static MMTK<VM>, tls: VMThread) {
-        // Spawn the controller thread.
-        let coordinator_worker = GCWorker::new(
-            mmtk,
-            usize::MAX,
-            self.clone(),
-            true,
-            self.coordinator_worker_shared.clone(),
-            deque::Worker::new_fifo(),
-        );
-        let gc_controller = GCController::new(
-            mmtk,
-            mmtk.gc_requester.clone(),
-            self.clone(),
-            coordinator_worker,
-        );
-        VM::VMCollection::spawn_gc_thread(tls, GCThreadContext::<VM>::Controller(gc_controller));
+        self.worker_group.create_workers(mmtk);
+        self.worker_group.spawn(tls)
+    }
+
+    /// Ask all GC workers to exit for forking.
+    pub fn stop_gc_threads_for_forking(self: &Arc<Self>) {
+        self.worker_group.prepare_surrender_buffer();
 
-        self.worker_group.spawn(mmtk, tls)
+        debug!("A mutator is requesting GC threads to stop for forking...");
+        self.worker_monitor.make_request(WorkerGoal::StopForFork);
+    }
+
+    /// Surrender the `GCWorker` struct of a GC worker when it exits.
+    pub fn surrender_gc_worker(&self, worker: Box<GCWorker<VM>>) {
+        let all_surrendered = self.worker_group.return_gc_worker_struct(worker);
```

Comment:
Obviously, that saves some allocation and initialization.  The native code may also save pointers to `GCWorker` instances as `VMWorkerThread`.  It is the safest and the most convenient to reuse the existing structures, although VMs still need to take care of setting thread-local variables (for PThread or C/C++'s `thread_local` variables).

Another difficulties is that we currently assume the number of GC threads is fixed.  We still have a fixed number of `WorkerShared` and `deque::Worker` instances created when the `WorkerGroup` is created.  We still need to reuse those data structures anyway, even if we destroy and recreate `GCWorker` instances.  That would be more complicated.

But one thing we can explore is the ability to dynamically scale the number of GC workers at run time.  In that way, we no longer assume a fixed number of workers, and we don't need to pre-create `WorkerShared` and related instances before spawning GC threads.  We can then simply scale the number of GC threads down to 0 before forking, and re-scale it up (maybe to a different number depending on how many processes the VM wishes to fork into) after forking.  That'll be another refactoring.

Reply:
> Obviously, that saves some allocation and initialization. 

I don't think allocation and initialization really matters here. Forking is heavy weight. Compared to that, creating a few structs for GC workers should be trivial.

> The native code may also save pointers to `GCWorker` instances as `VMWorkerThread`. It is the safest and the most convenient to reuse the existing structures, although VMs still need to take care of setting thread-local variables (for PThread or C/C++'s `thread_local` variables).

I don't think this is a valid reason either. After forking, all the GC threads exit. It is very unlikely that the binding somehow holds a reference to the GC threads and the GC workers.

> Another difficulties is that we currently assume the number of GC threads is fixed. We still have a fixed number of `WorkerShared` and `deque::Worker` instances created when the `WorkerGroup` is created. We still need to reuse those data structures anyway, even if we destroy and recreate `GCWorker` instances. That would be more complicated.

This is a valid point, based on the current code base. We create `WorkerGroup` before spawning GC workers. The reason is that plans need a reference to the scheduler, and the scheduler includes a worker group so we create the worker group before we create the plan. However, this is not necessary. We use `None` for the worker group when we create the scheduler, and only create the worker group when we spawn threads. In this way, we have to bear the cost of `unwrap` everytime we use worker group (which should be fine -- when a thread accesses worker group to modify other workers shared state, it is not expected to be cheap anyway). See https://github.com/qinsoon/mmtk-core/commit/1f495764f2502aba1b3c827fe9adb07960e7c244 for a prototype for this refactoring.

My concern is that there seems no valid point why we need the complexity to allow reusing the GC worker structs, and once we accept those changes, it may need extra efforts to remove it. I can submit a pul request to remove the pre-created `deque::Worker` list if you think that is reasonable.

Reply:
> I don't think allocation and initialization really matters here. Forking is heavy weight. Compared to that, creating a few structs for GC workers should be trivial.

That makes sense.  If that matters, it won't be feasible to scale the number of GC workers at run time, either.

> I don't think this is a valid reason either. After forking, all the GC threads exit. It is very unlikely that the binding somehow holds a reference to the GC threads and the GC workers.

Yes.  It should be unlikely that the binding keep references to one `GCWorker` from many different places.

Currently, the Ruby binding holds a pointer to the `GCWorker` instance in a thread-local variable for each GC worker thread.  Mutator threads are not aware of `GCWorker` instances.  The Ruby binding also creates GC workers as Rust threads (i.e. creating threads using the Rust standard library).  So the implementation is simple for Ruby.  The binding keeps track of how many GC worker threads actually exited (using `JoinHandle`), and the GC worker threads themselves can just exit.

In the OpenJDK binding, `MMTkCollectorThread` inherits `NamedThread`, and constructing that will need to register it to some global lists.  And `MMTkCollectorThread` holds a pointer to the `GCWorker` instance (in `MMTkCollectorThread::third_party_heap_collector`.  It is arguable that GC workers are reachable from the global list.  But I don't think that matters.  When a GC worker exits, the `MMTkCollectorThread` always needs to be removed from the global list and destructed, even if we keep the `GCWorker` instance.  OpenJDK doesn't need support for `fork()`, but I guess Android will be similar.  Android ART must have already taken care of this to support Zygote.

So I think you are right.  Well-written VMs and bindings shouldn't have pointers to `GCWorker` instances all over the place.

> This is a valid point, based on the current code base. We create `WorkerGroup` before spawning GC workers. The reason is that plans need a reference to the scheduler, and the scheduler includes a worker group so we create the worker group before we create the plan. However, this is not necessary. We use `None` for the worker group when we create the scheduler, and only create the worker group when we spawn threads. In this way, we have to bear the cost of `unwrap` everytime we use worker group (which should be fine -- when a thread accesses worker group to modify other workers shared state, it is not expected to be cheap anyway). See [qinsoon@1f49576](https://github.com/qinsoon/mmtk-core/commit/1f495764f2502aba1b3c827fe9adb07960e7c244) for a prototype for this refactoring.
> 
> My concern is that there seems no valid point why we need the complexity to allow reusing the GC worker structs, and once we accept those changes, it may need extra efforts to remove it. I can submit a pul request to remove the pre-created `deque::Worker` list if you think that is reasonable.

I think it is a good idea to make a dedicated pull request to refactor the creation of GC workers and remove the pre-created `deque::Worker`.  I looked at your refactoring.  I think the extra level of `unwrap()` is OK, but we should use safe code instead of relying on `Arc::to_raw` (casting `&` to `&mut` has undefined behavior anyway).

There are details that can be discussed.  For example, whether we use `Mutex<Option<WorkerGroup>>` or let a `WorkerGroup` have a `Mutex<Vec<Arc<GCWorkerShared>>>` instead.  The latter allows us to be aggressive and make the number of GC workers adjustable at run time.  But the former is also reasonable if we have multiple worker groups (e.g. one for concurrent GC and another for STW GC) so that a worker group may not always be available.

We can also move the `WorkerGroup` to a shared synchronized data structure, such as `WorkerMonitor` which I extensively used in this PR for synchronization.  One feature of this PR is that when the last GC worker parked, that worker will have already gained exclusive access to the `WorkerMonitor` struct.

My concern is that in the current `master` branch, there are a few complex states such as `enum WorkerGroupState { Sleeping, Working }` which I created for synchronizing between the workers and the coordinator.  It may be easier to do the refactoring after removing the coordinator.